### PR TITLE
Implement forgot my password functionality

### PR DESCRIPTION
**Pull request by OvidiuGi**

### CODE REVIEW

**Comment by ovidiuenache** [comment]
@OvidiuGi please change the name of the PR and mention what is implemented.

File: migrations/Version20220330145615.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+
+    }
+
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE messenger_messages CHANGE body body LONGTEXT NOT NULL COLLATE `utf8mb4_unicode_ci`, CHANGE headers headers LONGTEXT NOT NULL COLLATE `utf8mb4_unicode_ci`, CHANGE queue_name queue_name VARCHAR(255) NOT NULL COLLATE `utf8mb4_unicode_ci`');
```
Comment (ovidiuenache) [inline migrations/Version20220330145615.php:29]: Why did you remove an existing migration?

File: src/Controller/ForgotMyPasswordController.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        $this->router = $router;
+        $this->mailer = $mailer;
+    }
+
+    /**
+     * @Route(path="/forgot-password")
+     */
+    public function new(Request $request): Response
+    {
+        $task = new User();
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:45]: What do you think about the name of this variable?

File: src/Controller/ForgotMyPasswordController.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+        $form = $this->createForm(ForgotMyPasswordType::class);
+
+        $form->handleRequest($request);
+        if ($form->isSubmitted() && $form->isValid()) {
+            $givenEmail = $form->getData()['email'];
+            $userRepository = $this->entityManager->getRepository(User::class);
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:52]: The repo is a service and can be injected in the controller

File: src/Controller/ForgotMyPasswordController.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+            $user = $userRepository->findOneBy(['email' => $givenEmail]);
+
+            if ($user) {
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:56]: Always write explicit expressions

File: src/Controller/ForgotMyPasswordController.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                    'reset_password',
+                    ['token' => $user->forgotPasswordToken],
+                    UrlGeneratorInterface::ABSOLUTE_URL
+                );
+                $email = (new Email())
+                    ->from('gireadaovidiu123@gmail.com')
+                    ->to($givenEmail)
+                    ->subject('Change password for account')
+                    ->text('Click the link to change password')
+                    ->html("<a href=$newPasswordUrl>Click me!</a>");
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:73]: Use a twig if you want to get fancy

File: src/Controller/ForgotMyPasswordController.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    public function reset(Request $request)
+    {
+        $token = $request->query->all()['token'];
+
+        $form = $this->createForm(ResetPasswordType::class);
+        $userRepository = $this->entityManager->getRepository(User::class);
+
+        $user = $userRepository->findOneBy(['forgotPasswordToken' => $token]);
+        $form->handleRequest($request);
+        if ($user) {
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:96]: same

File: src/Controller/ForgotMyPasswordController.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     * @Route(path="/reset-password", name="reset_password")
+     */
+    public function reset(Request $request)
+    {
+        $token = $request->query->all()['token'];
+
+        $form = $this->createForm(ResetPasswordType::class);
+        $userRepository = $this->entityManager->getRepository(User::class);
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:92]: same

**Review by ovidiuenache — CHANGES_REQUESTED** [review]
Please address the feedback above

File: src/Controller/ForgotMyPasswordController.php (hunk 3 above)
Comment (OvidiuGi) [reply src/Controller/ForgotMyPasswordController.php:52]: I worked with a version of main without repository. I will change it

File: src/Controller/ForgotMyPasswordController.php (hunk 2 above)
Comment (OvidiuGi) [reply src/Controller/ForgotMyPasswordController.php]: Renamed it.

File: src/Controller/ForgotMyPasswordController.php (hunk 5 above)
Comment (OvidiuGi) [reply src/Controller/ForgotMyPasswordController.php]: Maybe later..

File: migrations/Version20220330145615.php (hunk 1 above)
Comment (OvidiuGi) [reply migrations/Version20220330145615.php]: I dropped the database and deleted the migrations from local to make room for new ones..

File: migrations/Version20220331152457.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+{
+    public function getDescription(): string
+    {
+        return '';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('CREATE TABLE building (id INT AUTO_INCREMENT NOT NULL, start_time DATETIME NOT NULL, end_time DATETIME NOT NULL, PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
```
Comment (ovidiuenache) [inline migrations/Version20220331152457.php:23]: Why did you remove the existing migration and create one with everything?

File: src/Controller/ForgotMyPasswordController.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            $this->logger->info('No user found for token:' . $request->query->all()['token']);
+
+            return new Response('The token is not valid', Response::HTTP_NOT_FOUND);
+        }
+        $form = $this->createForm(ResetPasswordType::class);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            if (date_diff(new \DateTime('now'), $user->getForgotPasswordTokenTime())->i > 60) {
+                return new Response('The link expired', Response::HTTP_METHOD_NOT_ALLOWED);
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:93]: Wrong HTTP return code

File: src/Controller/ForgotMyPasswordController.php [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * @Route(path="/reset-password", name="reset_password")
+     */
+    public function reset(Request $request): Response
+    {
+        $user = $this->userRepository->findOneBy(['forgotPasswordToken' => $request->query->all()['token']]);
+        if (null === $user) {
+            $this->logger->info('No user found for token:' . $request->query->all()['token']);
+
+            return new Response('The token is not valid', Response::HTTP_NOT_FOUND);
```
Comment (ovidiuenache) [inline src/Controller/ForgotMyPasswordController.php:86]: Wrong HTTP return code

**Review by ovidiuenache — CHANGES_REQUESTED** [review]
Please address the feedback above

File: migrations/Version20220331152457.php (hunk 8 above)
Comment (OvidiuGi) [reply migrations/Version20220331152457.php:23]: I had an error when i wanted to add a field to user table with a migration, so i dropped the database and recreated it.

File: src/Controller/ForgotMyPasswordController.php (hunk 9 above)
Comment (OvidiuGi) [reply src/Controller/ForgotMyPasswordController.php:93]: Changed with 401

File: src/Controller/ForgotMyPasswordController.php (hunk 10 above)
Comment (OvidiuGi) [reply src/Controller/ForgotMyPasswordController.php:86]: Changed with 401.

**Review by ovidiuenache — APPROVED** [review]

**Comment by ovidiuenache** [comment]
@OvidiuGi approved
